zebra-consensus/src/transaction/tests.rs

```diff
@@ -730,7 +730,8 @@ fn v4_with_sapling_spends() {
             })
             .await;
 
-        assert_eq!(result, Ok(expected_hash));
+        //assert_eq!(result, Ok(expected_hash));
+        assert_eq!(result, Err(TransactionError::InvalidValueBalance));
```

Comment:
It looks like a bug in the `remaining_transaction_value` method.

The transaction of the test haves 1 output of amount `25500000000` and no inputs. Then it haves a sapling value of `25500010000`.

Our rule checks the following is non negative:

```
(self.transparent - (self.sprout + self.sapling + self.orchard)?)?
```

But in our case the rule is broken:

(0 - 25500000000) - (0 + 25500010000 + 0) = -51000010000

Reply:
It looks that the rule needs to be:

```
(self.transparent + (self.sprout + self.sapling + self.orchard)?)?
```

However not sure.

Reply:
It looks like we've got a sign wrong somewhere - we need to swap a positive to a negative.

There are a few places where we need to check signs.

Low-level methods:
* `transparent::Input::value_balance`
* `transparent::Output::value_balance`
* `JoinSplitData::value_balance`
* `sapling::ShieldedData::value_balance`
* `orchard::ShieldedData::value_balance`

Transaction methods:
* `Transaction::transparent_value_pool`
* `Transaction::sprout_value_pool`
* `Transaction::sapling_value_pool`
* `Transaction::orchard_value_pool`
* `Transaction::value_balance`

Value balance methods:
* `ValueBalance::remaining_transaction_value`

Which ones don't match the definitions in the design?
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0012-value-pools.md#definitions

While you're doing this, it might help to copy the matching definitions from the design to the methods. That way, reviewers can easily check the design against the code.

Reply:
Oh, and if you have a choice, change the sign in the lowest level method possible. That way, you're not swapping signs low down, then changing them again higher up!

Reply:
Thank you for this and sorry because i didn't paid enough attention to the definitions.

The problem here was that the sapling and orchard value balances are the negation of the sum:

`sapling value balance` - The change in the sapling value pool. The negation of the sum of all `valueBalanceSapling` fields.
`orchard value balance` - The change in the orchard value pool. The negation of the sum of all `valueBalanceOrchard` fields.

This was fixed in https://github.com/ZcashFoundation/zebra/pull/2486/commits/d0053fdf98674b6109c26caed81294733d9fab41

More documentation was also added.